upf/pfcp.c

```diff
@@ -781,8 +781,21 @@ format_f_teid (u8 * s, va_list * args)
 		format_ip6_address, &v->ip6);
   else if ((v->flags & 0xf) == F_TEID_CH)
     s = format (s, "%d,CH:1", v->teid);
+  else if ((v->flags & 0xf) == (F_TEID_CH | F_TEID_V4))
+    s = format (s, "%d,V4:1,CH:1", v->teid);
+  else if ((v->flags & 0xf) == (F_TEID_CH | F_TEID_V6))
+    s = format (s, "%d,V6:1,CH:1", v->teid);
+  else if ((v->flags & 0xf) == (F_TEID_CH | F_TEID_V4 | F_TEID_V6))
+    s = format (s, "%d,V4:1,V6:1,CH:1", v->teid);
   else if ((v->flags & 0xf) == (F_TEID_CH | F_TEID_CHID))
     s = format (s, "%d,CH:1,CHID:%d", v->teid, v->choose_id);
+  else if ((v->flags & 0xf) == (F_TEID_CH | F_TEID_CHID | F_TEID_V4))
+    s = format (s, "%d,V4:1,CH:1,CHID:%d", v->teid, v->choose_id);
+  else if ((v->flags & 0xf) == (F_TEID_CH | F_TEID_CHID | F_TEID_V6))
+    s = format (s, "%d,V6:1,CH:1,CHID:%d", v->teid, v->choose_id);
+  else if ((v->flags & 0xf) == (F_TEID_CH | F_TEID_CHID | F_TEID_V4 |
+				F_TEID_V6))
+    s = format (s, "%d,V4:1,V6:1,CH:1,CHID:%d", v->teid, v->choose_id);
```

Comment:
This part could be rewritten to check each flag in turn separately, adding corresponding parts like `V4:1` etc one by one

Reply:
you could even do it in one go using the `!!` trick:

```
format (s, "%d,V4:%d,V6:%d",
    !!(v->flags & F_TEID_V4), !!(v->flags & F_TEID_V6),!!(v->flags & F_TEID_CH))
```

You could also first print all flags and then add TEID, CH-Id and the IPs to the end if the flags are set. That would be slightly more verbose than the old format, but would work.

Reply:
Refactored this debug function